### ISSUE TITLE
Long term stability observations

### ISSUE THREAD

**Issue by ceresek** [dacapobench/dacapobench#269]
Hello, we've been looking at long-term stability of the benchmarks and noticed a few patterns that might be of interest. I'm attaching plots with execution time on X axis (8 hours), iteration time on left Y axis in color (each color is one independent benchmark execution), heap size after forced GC on right Y axis in gray (0 based).

![h2o](https://github.com/dacapobench/dacapobench/assets/6158703/2916abcf-e4fd-49ad-822f-fab24b8dd505)
![jme](https://github.com/dacapobench/dacapobench/assets/6158703/ff3c2055-d6ca-4a0c-95bd-fac938816649)
![jython](https://github.com/dacapobench/dacapobench/assets/6158703/cf266cd8-6337-4b57-b2d0-21e46e26afb9)
![kafka](https://github.com/dacapobench/dacapobench/assets/6158703/61a78c3a-122d-4b1e-865b-cdeac43834c9)
![tomcat](https://github.com/dacapobench/dacapobench/assets/6158703/78171bed-601b-428d-a68e-36f63781efb0)

All data was collected on recent Linux with OpenJDK 11 and no special settings apart from `-Xms` and `-Xmx`. In all cases there seems to be a gradual heap size growth and sometimes gradual slow down (sometimes not really significant though), which may be a behavior that was not intended. I'm not sure if these are of interest to you, it would be nice to know if you observe similar behavior.

**Comment by steveblackburn** [comment]
Thanks for this.

It is quite interesting.  A quick look at the nominal stats shows that [h2o](https://github.com/dacapobench/dacapobench/blob/main/benchmarks/bms/h2o/stats-nominal.yml#L11) and [jython](https://github.com/dacapobench/dacapobench/blob/main/benchmarks/bms/jython/stats-nominal.yml#L18) are already established as being among the highest memory leakers (3rd and 5th worst).

The `GLK` nominal statistic measures heap growth over ten iterations by measuring the minimum heap size for G1 for 1 iteration and then comparing that with the minimum heap size for 10 iterations.   The others you list have `GLK` values of `0`, which of course does not mean that they don't leak, but if they do leak, the rate of leakage was not sufficient to be observed with this very simple metric.

From the standpoint of the benchmark suite, I'm very interested to know whether you have any insight as to whether there's a problem with the _suite_ or whether these behaviors are simply artefacts of the upstream codebase/workload.  

If the former, then of course I'd like to address them, ASAP.

If the latter, this is 'fine' from the point of view of the suite, since part of our mission is to present 'real-world' workloads 'warts and all', so if they have interesting performance pathologies, then that's not in itself a bad thing because as long as we're confident that it's not due to DaCapo itself, it is reflective of the sorts of pathologies that appear in real-world systems, and is thus a worthwhile subject of study.

If you have any info on the reasons for this behavior, please do create issues as appropriate.

**Comment by steveblackburn** [comment]
Some further comments:

First, there appear to be both black and grey markers.  Is that just an illusion?  If not, what do the grey markers indicate? (You only mentioned black, I think).

### H2O

This is the most puzzling, and given our past experience, the least surprising.  h2o is something of a monster of a workload with lots of strange built-in features that may confound performance analysis.   It would be very interesting to understand the performance slowdowns, and the curious bifurcation (very clear on the right, where performance results appear to be bimodal).

### JME

This is interesting for two reasons.   First the performance slowdown closely tracks the space leak (coincidence, perhaps).  Second, each of the trials shows the same slope, but has distinctly different performance.  This is curious.   The total slowdown of the orange trail is not as large as the difference between it and the magenta trial.  Why would the trials be so different, I wonder?   Note that JME spends a large amount of its time on the GPU.

### Jython

There's a slow memory leak, which would be nice to understand.  But the performance slowdown here is really interesting, and unlike JME, it seems unlikely to be attributable to the leak (I might be wrong).  I can imagine that the early step speedup is due to JIT behavior, but it is much less clear why performance would then degrade by roughly 20-25%.  I'd love to know more.

### Kafka

The kafka results seem very stable, with a slow memory leak.  Since the workload is being [substantially re-worked](https://github.com/dacapobench/dacapobench/pull/268), I won't invest too much time looking further into this.

### Tomcat

My read is that this data suggests that tomcat is very stable, with a very small memory leak.  The performance results show a curious sinusoidal pattern which is replicated across each of the trials.  I have no idea what this is, but note that the amplitude of the sinusoid is similar to the variance among iterations, and is smaller than the variance between trails (ie it is small).  Three questions that come to mind:

1. What is the source of the sinusoidal pattern and why are all trials so perfectly in phase?
2. Why are each of the trails so discretely different (albeit by only small amounts)?
3. What is the source of the (tiny) memory leak?

I'm not expecting you to provide answers to any of these, but offering them as an indication of what I find interesting.

**Comment by caizixian** [comment]
Yep, I can reproduce the jython slowdown using Temurin 21 and 2x heap (on i9-9900K)

![image](https://github.com/dacapobench/dacapobench/assets/2891235/486510d4-5809-44d1-bedc-9f5b66f72c2e)

**Comment by ceresek** [comment]
Thank you for your insights ! I very much understand your position wrt "suite" vs "framework" issue, and if we find out more about any of these behaviors, I'll leave a note (at the same time, I must add we are not investigating this as a prime target, the observations are a side product of other benchmarking work).

In the meantime, I can offer some clarification on the points you have raised above:

- Re black and gray markers: We have memory usage info for multiple runs, but the plots already use colors for distinguishing timing, doing the same for memory usage would make the plots a mess. We therefore plot memory usage of all runs in black, but with 10% alpha channel, so values common to more runs look more "black" and values unique for some runs are more "gray". Sorry for the confusion.

- Re h2o clusters: Similar bimodal behavior is visible on the metric tracking young collection count per iteration. Other metrics we collect (e.g. dynamic instruction count or memory references per iteration) track time quite closely, suggesting the workload really performs gradually more work per iteration. The slower of the two clusters is much less common (about an order of magnitude so), the pattern is not regular (at a glance I can see anywhere between 5-30 fast iterations between two slow iterations, but also two slow iterations back to back).

- Re jme: The similarity of slopes between time and memory usage is I think only an artifact of graph scaling. What is interesting is that while the iteration time is constantly growing (except for sudden mode changes that shift the absolute value but not the slope), the instruction count per iteration remains stable in most iterations, and only a small share of iterations executes gradually more instructions. Memory accesses (in terms of last level cache references and misses) are also stable.

- Re jython: I would agree that the initial jump can be explained by JIT activity. The gradual slowdown is again faithfully tracked by other counters (instructions, cache references, cache misses, and a few more), some changes are not large in absolute terms (across the range where the iteration time grows from 3.2s to 4.2s, cache references go from ~240M to ~420M but cache misses only from ~102M to ~111M). Over the same range, memory usage grows from ~3.9 GB to ~4.2 GB, very roughly 37 kB per iteration.

- Re tomcat: The JIT activity of this benchmark is interesting. Where other workloads gradually cease compiling, and typically spend anywhere between ~10s to ~200s of processor time compiling over the 8 hour execution, this workload keeps compiling with no pause, consuming ~1000s of processor time compiling over the 8 hour execution (no other workload comes close, JIT in the other worst offenders is an order of magnitude less active). The memory usage grows at around 95 kB per iteration, as always some of this may well be storing the measurements.

Sorry for not having deeper insights, this is just commenting on the data we have (mostly bits that cannot be seen from the few pictures uploaded).
